File: day2/main2.py
with open("input") as f:
    ans = 0
    lines = [x.rstrip() for x in f.readlines()]
    for idx, line in enumerate(lines):
        subsets = [x.split(',') for x in line.split(':')[1].split(';')]
        num_color = {'red':0, 'green':0, 'blue':0}
        for subset in subsets:
            for color in subset:
                cpt = int(color.split()[0])
                c = color.split()[1]
                num_color[c] = max(cpt, num_color[c])
            power=1
        for x in num_color.values():
            power *= x
        ans += power
    print(ans)

File: day2/main2.cpp
#include <bits/stdc++.h>
using namespace std;

int main()
{
  ifstream ifs("input");
  string line;
  int ans=0;
  while(getline(ifs, line)){
    unordered_map<string, int> num_color;
    stringstream ssline(line);
    string game, semicolon, subset;
    int n;
    ssline >> game >> n >> semicolon;
    while(getline(ssline, subset, ';')){
      stringstream ss(subset);
      string color;
      while(getline(ss, color, ',')){
        string c;
        int cpt;
        stringstream sscolor(color);
        sscolor >> cpt >> c;
        num_color[c] = max(num_color[c],cpt);
      }
    }
    int power = 1;
    for(auto& l:num_color){
      power*=l.second;
    }
    ans+=power;
  }
  cout << ans << endl;
}
